Gutenpack: Only load frontend assets as required

enejb opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

Fixes #10325 

Requires you to build the blocks via this branch:
https://github.com/Automattic/wp-calypso/pull/28066 

This PR implement a new `jetpack_register_block()` function. Which helps Jetpack have a better understand what blocks are supported and active. Different modules could be disabled. This information could then also be used in the editor to show/hide the blocks there as well.

I only added the tiled-gallery block because for now it is the only block that has frontend assets. 

#### Testing instructions:
URL: [JN test](https://jurassic.ninja/create/?jetpack-beta&branch=update/jetpack-gutenber-only-load-assets-as-required&shortlived&wp-debug-log&gutenberg&gutenpack)

Create a new tile gallery block. 
Notice that the tile gallery view.js is only loaded on the page where the block appears.


#### Proposed changelog entry for your changes:
Only load front end assets as required by the block.

enejb committed:
Gutenblock: Add the frontend code only if the block is being called.

This Removed the loading of the front end code that comes as a bundle of all the files.
And make sure that it loads only the front end code that is needed when the block is going to be displayed.

enejb committed:
Register the tiled-galley code block

This lets us register the tiled gallery code to be loaded on the front end only

enejb committed:
Changes applied to #10397 by gititon

Props @gititon.

enejb committed:
Minor bug fix

jeherve commented on class.jetpack-gutenberg.php:
```diff
@@ -1,78 +1,71 @@
 <?php
+function jetpack_register_block( $type, $args = array() ) {
```
Could you add a docblock for that function? I suspect we'll get to use that function a lot in the future, so it would be useful and helpful if it was fully documented so our IDEs can use that information.

I would recommend doing the same with other functions and variable declarations in this file to make our linters happy.

jeherve commented on class.jetpack.php:
```diff
@@ -537,7 +537,7 @@ private function __construct() {
 		 */
 		require_once JETPACK__PLUGIN_DIR . 'class.jetpack-gutenberg.php';
-		add_action( 'enqueue_block_assets', array( 'Jetpack_Gutenberg', 'enqueue_block_assets' ) );
+		add_action( 'init', array( 'Jetpack_Gutenberg', 'load_blocks' ), 1000 );
```
Why the big priority here?

enejb replied:
The reason for it is to make sure that modules are loaded before the blocks are registered. 
the default value should be fine since modules are loaded before init anyways.

jeherve commented on class.jetpack-gutenberg.php:
```diff
@@ -1,78 +1,71 @@
-				: JETPACK__VERSION;
+	public static function load_assets_as_required( $type ) {
+		// Enqueue styles
+		$style_relative_path = '_inc/blocks/'. $type .'/view.'. ( is_rtl() ? '.rtl' : '' ) .'css';
```
I think you are missing some spaces around the concat operators here.

jeherve replied:
I think there is an issue with the `.` for the rtl file here. You're going to end up with a path like `_inc/blocks/tiled-gallery/view..rtlcss`

jeherve commented on class.jetpack-gutenberg.php:
```diff
@@ -1,78 +1,71 @@
+		$style_relative_path = '_inc/blocks/'. $type .'/view.'. ( is_rtl() ? '.rtl' : '' ) .'css';
+		if ( self::block_has_asset( $style_relative_path ) ) {
+			$style_version       = self::get_asset_version( $style_relative_path );
+			$view_style  = plugins_url( $style_relative_path, JETPACK__PLUGIN_FILE );
```
`=` signs are not aligned here.

jeherve commented on class.jetpack-gutenberg.php:
```diff
@@ -1,78 +1,71 @@
-			$version
-		);
+		// Enqueue script
+		$script_relative_path = '_inc/blocks/'. $type .'/view.js';
```
Missing spaces here as well.

jeherve commented on modules/tiled-gallery.php:
```diff
@@ -30,3 +30,12 @@ function jetpack_tiled_gallery_configuration_load() {
+jetpack_register_block( 'tiled-gallery', array(
+	'render_callback' => 'jetpack_tiled_gallery_load_assets'
+) );
+function jetpack_tiled_gallery_load_assets( $attr, $content ) {
```
Do you think you could add a docblock here so we know what the parameters refer to?

jeherve reviewed (dismissed):
I don't seem to see any tiled gallery on the front end of my site when inserting a gallery via the Gutenberg block. I consequently do not see the gallery assets in the source code either. Maybe that's partly because #10375 has not been merged yet?

I also left a few comments about minor coding standards changes you could make here, so our code editors don't shout at us when we open the file :)

There also seems to be a small issue with how the file is built for RTL sites.

lezama committed:
fix linting

lezama committed:
missing !

lezama commented:
@jeherve thanks for the review!
@enejb I applied linting and added a missing `!` now we are enqueuing the assets 🎉

jeherve commented:
@lezama Thanks! We still have that RTL issue though:
https://github.com/Automattic/jetpack/pull/10405#discussion_r228441785

And the missing docblocks. :)

enejb committed:
Added code dock blocks and set the default priority

enejb commented:
@jeherve Can you elaborate on the RTL issue? I didn't find any issues but I also wasn't able to test it properly since the files are not there.

enejb committed:
Fixed the RTL issue

enejb commented:
Nevermind good catch on the RTL issue. 🤦‍♂️

jeffersonrabb commented:
When testing locally (JP docker) I'm getting `This block contains unexpected or invalid content.` for the Tiled Gallery after saving, and nothing displays in the frontend after publishing the post. The JN link is also failing. However, I do see the tiled-gallery's `view.js` file loading in the frontend, and I don't see it if the block is removed, so that's looking good.

simison commented:
The tiled gallery currently in `master` isn't very stable indeed (and the new one isn't merged yet https://github.com/Automattic/wp-calypso/pull/27458).

It might make sense to update our testing block `hello-dolly` to contain some styles and use that for testing... you could probably re-use this PR: https://github.com/Automattic/wp-calypso/pull/27241

tyxla commented on class.jetpack-gutenberg.php:
```diff
@@ -16,63 +54,56 @@ public static function is_gutenberg_available() {
+		return (bool) apply_filters( 'jetpack_gutenberg', true );
+	}
+
+	public static function load_assets_as_required( $type ) {
```
Do we want to make sure that the `$type` is slug-like before using it? Perhaps passing it to `sanitize_title_with_dashes()` or `sanitize_title()` should be enough.

enejb replied:
I think we are OK with not so much checking since this is only supposed to be used from withing code. And I expect code review to be able to catch such things as well.

tyxla replied:
I agree, but also any third party can call `jetpack_register_block()` or `Jetpack_Gutenberg::add_block()`, right? Never can be too safe 😉 

Anyway, I'll leave this to your judgement.

tyxla commented on class.jetpack.php:
```diff
@@ -537,7 +537,7 @@ private function __construct() {
 		 * Prepare Gutenberg Editor functionality
 		 */
 		require_once JETPACK__PLUGIN_DIR . 'class.jetpack-gutenberg.php';
-		add_action( 'enqueue_block_assets', array( 'Jetpack_Gutenberg', 'enqueue_block_assets' ) );
```
Seems like we'll no longer use core Gutenberg's hook `enqueue_block_assets` to enqueue the assets. Core Gutenberg recommends using that hook, because it assures that its necessary assets [will already be loaded](https://github.com/WordPress/gutenberg/blob/3b66a7f21f646a4a78379330f3ae4181ccefeac0/lib/client-assets.php#L1152). Do we want to use it after all, even if we use `init` for our business logic?

enejb replied:
I think we are fine with not using it. since the current enqueue_block_assets hood also fires on the front end and results in assets being loaded when they don't have to be.

tyxla commented on modules/tiled-gallery.php:
```diff
@@ -30,3 +30,21 @@ function jetpack_tiled_gallery_configuration_load() {
+ *
+ * @return mixed
+ */
+function jetpack_tiled_gallery_load_assets( $attr, $content ) {
```
It doesn't seem that the attributes passed here are used for anything.

tyxla commented on modules/tiled-gallery.php:
```diff
@@ -30,3 +30,21 @@ function jetpack_tiled_gallery_configuration_load() {
+ * @param $attr array - array of attributes
+ * @param $content string - content block
+ *
+ * @return mixed
```
It appears that we're returning the original `$content`, and nothing is modifying it, so I'd expect this to still be a string.

tyxla commented on modules/tiled-gallery.php:
```diff
@@ -30,3 +30,21 @@ function jetpack_tiled_gallery_configuration_load() {
+
+// Tile-gallery block definition can be found in wp-calypso repo
+jetpack_register_block( 'tiled-gallery', array(
+	'render_callback' => 'jetpack_tiled_gallery_load_assets'
```
It's worth explaining why we pass a function that loads assets as a render callback - it can be confusing without reading the function body.

enejb commented on class.jetpack-gutenberg.php:
```diff
@@ -16,63 +54,56 @@ public static function is_gutenberg_available() {
-		 * @param bool false Whether to load Gutenberg blocks from CDN
+		 * @param bool true Whether to load Gutenberg blocks
 		 */
-		if ( apply_filters( 'jetpack_gutenberg_cdn', false ) ) {
```
No there is still the CDN option but only for the editor assets. I didn't want to remove it since it was already taken care of by a different PR.

enejb commented on class.jetpack-gutenberg.php:
```diff
@@ -16,63 +54,56 @@ public static function is_gutenberg_available() {
+		return (bool) apply_filters( 'jetpack_gutenberg', true );
+	}
+
+	public static function load_assets_as_required( $type ) {
```
True. But do we want those filters right now? I think they can be done in a different PR when they are requested. There is also the previous way of dequeuing the assets. Which is done just before enqueuing them.

enejb committed:
Minor doc block fixes

jeherve committed:
Gutenberg: fix all linter errors in file.

We'll rely on those functions a lot in the next few months, the different comments should help us get all
the info we need in our IDEs.
It also acts as a good example for other folks who may be interested in how Jetpack adds support for its
blocks.

jeherve commented on class.jetpack-gutenberg.php:
```diff
@@ -16,63 +78,77 @@ public static function is_gutenberg_available() {
-			$version
-		);
+		// Enqueue script.
+		$script_relative_path = '_inc/blocks/' . $type . '/view.js';
```
Since the blocks now live in subdirectories, we will need to update this accordingly:
https://github.com/Automattic/jetpack/blob/20da538a07f111c238f50311e2f27a1605f7c930/gulpfile.babel.js#L255

jeherve replied:
Done in 54940f3cc00f07a1e61ab8b4298fe90ac8cca07e

simison commented:
Gulp task that extracts strings for translations from block JS files should be updated to include sub-folders:

https://github.com/Automattic/jetpack/blob/20da538a07f111c238f50311e2f27a1605f7c930/gulpfile.babel.js#L255-L259

jeherve committed:
Language extraction: account for blocks in sub-directories.

@see https://github.com/Automattic/jetpack/pull/10405#discussion_r229280595

jeherve commented:
Rebased and updated to account for https://github.com/Automattic/jetpack/pull/10405#discussion_r229280595

It's hard to test with just Tiled Galleries, though.

enejb committed:
Sanitize the type.

enejb committed:
Sanitize the type again

oskosk reviewed (approved):
LGTM!

oskosk commented:
Testing this PR on WordPress 5.0 Beta 2 (Gutenberg in core, not as a plugin) I get this error when attempting to include the Gallery block

`react-dom.min.js?ver=5.0-beta2-43845:104 TypeError: Cannot read property 'length' of undefined`

![image](https://user-images.githubusercontent.com/746152/47752474-4e6e4780-dc73-11e8-9261-8ed51f8c83b2.png)

jeherve commented:
Noting that #10491 reverted some of our work here.
